c7n/resources/sns.py

```diff
@@ -77,6 +77,96 @@ def format_resource(self, r):
         return envelope
 
 
+@SNS.filter_registry.register('has-statement')
+class HasStatementFilter(Filter):
+    """Find SNS topics with set of matching access policy statements.
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: sns-check-statement-id
+                resource: sns
+                filters:
+                  - type: has-statement
+                    statement_ids:
+                      - BlockNonSSL
+
+            policies:
+              - name: sns-check-block-non-ssl
+                resource: sns
+                filters:
+                  - type: has-statement
+                    statements:
+                      - Effect: Deny
+                        Action: 'SNS:Publish'
+                        Principal: '*'
+                        Condition:
+                            Bool:
+                                "aws:SecureTransport": "false"
+    """
+    schema = type_schema(
+        'has-statement',
+        statement_ids={'type': 'array', 'items': {'type': 'string'}},
+        statements={
+            'type': 'array',
+            'items': {
+                'type': 'object',
+                'properties': {
+                    'Sid': {'type': 'string'},
+                    'Effect': {'type': 'string', 'enum': ['Allow', 'Deny']},
+                    'Principal': {'anyOf': [
+                        {'type': 'string'},
+                        {'type': 'object'}, {'type': 'array'}]},
+                    'NotPrincipal': {
+                        'anyOf': [{'type': 'object'}, {'type': 'array'}]},
+                    'Action': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'NotAction': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'Resource': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'NotResource': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'Condition': {'type': 'object'}
+                },
+                'required': ['Effect']
+            }
+        })
+
+    def process(self, topics, event=None):
+        return list(filter(None, map(self.process_topic, topics)))
+
+    def process_topic(self, t):
+        p = t.get('Policy')
+        if p is None:
+            return None
+        p = json.loads(p)
+
+        required = list(self.data.get('statement_ids', []))
+        statements = p.get('Statement', [])
+        for s in list(statements):
+            if s.get('Sid') in required:
+                required.remove(s['Sid'])
+
+        required_statements = list(self.data.get('statements', []))
```

Comment:
maybe we should also have a `strict` param to state that the statement should match all the params in the custodian policy and not have any additional attributes, e.g. extra conditions

Reply:
@thisisshi I think @StevenGunn is just following the existing code that we already have here https://github.com/cloud-custodian/cloud-custodian/blob/master/c7n/resources/s3.py#L814-L900. If we want to have a different behavior, we would also need to be consistent and update the existing code. My opinion is we shouldn't change the existing behavior of how we handle policy statement.

@StevenGunn Talking about existing code, instead of copying and pasting the same code into the SNS resource, we should create a dedicated filter class for it. That way, we can resuse the code across any resources that have policy statements. Here's what it would look like:

`filters/policystatement.py`
```
# Copyright The Cloud Custodian Authors.
# SPDX-License-Identifier: Apache-2.0
import json

from .core import Filter
from c7n.utils import type_schema, format_string_values

class HasStatementFilter(Filter):
    """Find resources with matching access policy statements.
    :example:
    .. code-block:: yaml
            policies:
              - name: sns-check-statement-id
                resource: sns
                filters:
                  - type: has-statement
                    statement_ids:
                      - BlockNonSSL
            policies:
              - name: sns-check-block-non-ssl
                resource: sns
                filters:
                  - type: has-statement
                    statements:
                      - Effect: Deny
                        Action: 'SNS:Publish'
                        Principal: '*'
                        Condition:
                            Bool:
                                "aws:SecureTransport": "false"
    """
    schema = type_schema(
        'has-statement',
        statement_ids={'type': 'array', 'items': {'type': 'string'}},
        statements={
            'type': 'array',
            'items': {
                'type': 'object',
                'properties': {
                    'Sid': {'type': 'string'},
                    'Effect': {'type': 'string', 'enum': ['Allow', 'Deny']},
                    'Principal': {'anyOf': [
                        {'type': 'string'},
                        {'type': 'object'}, {'type': 'array'}]},
                    'NotPrincipal': {
                        'anyOf': [{'type': 'object'}, {'type': 'array'}]},
                    'Action': {
                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
                    'NotAction': {
                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
                    'Resource': {
                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
                    'NotResource': {
                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
                    'Condition': {'type': 'object'}
                },
                'required': ['Effect']
            }
        })

    def process(self, resources, event=None):
        return list(filter(None, map(self.process_resource, resources)))

    def process_resource(self, resource):
        p = resource.get('Policy')
        if p is None:
            return None
        p = json.loads(p)

        required = list(self.data.get('statement_ids', []))
        statements = p.get('Statement', [])
        for s in list(statements):
            if s.get('Sid') in required:
                required.remove(s['Sid'])

        required_statements = list(self.data.get('statements', []))

        required_statements = format_string_values(list(self.data.get('statements', [])),
                                                   **self.get_std_format_args(resource))

        for required_statement in required_statements:
            for statement in statements:
                found = 0
                for key, value in required_statement.items():
                    if key in statement and value == statement[key]:
                        found += 1
                if found and found == len(required_statement):
                    required_statements.remove(required_statement)
                    break

        if (self.data.get('statement_ids', []) and not required) or \
           (self.data.get('statements', []) and not required_statements):
            return resource
        return None

    def get_std_format_args(self, _):
        return {
            'account_id': self.manager.config.account_id,
            'region': self.manager.config.region,
        }
```

`resources/s3.py` can then be updated with
```
...
import c7n.filters.policystatement as polstmt_filter
...
@S3.filter_registry.register('has-statement')
class HasStatementFilter(polstmt_filter.HasStatementFilter, BucketFilterBase):
    pass
```

and `resources/sns.py` can be updated with
```
...
import c7n.filters.policystatement as polstmt_filter
...
@SNS.filter_registry.register('has-statement')
class HasStatementFilter(polstmt_filter.HasStatementFilter):
    def get_std_format_args(self, topic):
        return {
            'topic_arn': topic['TopicArn']
        }
```

Reply:
agreed and makes sense, thank you @darrendao. Will be making the reusability enhancements.